pkg/engine/virtualmachinescalesets.go

```diff
@@ -828,3 +815,36 @@ func addCustomTagsToVMScaleSets(tags map[string]string, vm *compute.VirtualMachi
 		}
 	}
 }
+
+func associateAddonIdentitiesToVMSS(addonProfiles map[string]api.AddonProfile, virtualMachineScaleSet *compute.VirtualMachineScaleSet) {
+	if virtualMachineScaleSet == nil {
+		return
+	}
+	for _, addonProfile := range addonProfiles {
+		if addonProfile.Enabled && addonProfile.Identity != nil && addonProfile.Identity.ResourceID != "" {
+			// We need to associate addon's identity to VMSS, there're 3 cases:
+			// 1. virtualMachineScaleSet.Identity is nil. In this case, we need to initialize "virtualMachineScaleSet.Identity" and set its type to UserAssigned.
+			// 2. virtualMachineScaleSet.Identity is not nil, and its type is SystemAssigned. This case will happen in an MSI cluster and the VMSS uses system
+			// assigned identity. In this case, we need to set current `virtualMachineScaleSet.Identity.Type` to `ResourceIdentityTypeSystemAssignedUserAssigned`.
+			// 3. virtualMachineScaleSet.Identity is not nil, and its type is UserAssigned. This case will happen in an MSI cluster and the VMSS uses user assigned
+			// identity. In this case, no additional step is needed. Just keep `virtualMachineScaleSet.Identity.Type` unchanged and fill in addon's identity later.
+			// Note: virtualMachineScaleSet.Identity is not nil and its type is None will NEVER happen in current AKS-Engine's implementation.
+			if virtualMachineScaleSet.Identity == nil {
+				virtualMachineScaleSet.Identity = &compute.VirtualMachineScaleSetIdentity{
+					Type:                   compute.ResourceIdentityTypeUserAssigned,
+					UserAssignedIdentities: make(map[string]*compute.VirtualMachineScaleSetIdentityUserAssignedIdentitiesValue),
+				}
+			} else if virtualMachineScaleSet.Identity.Type == compute.ResourceIdentityTypeSystemAssigned {
+				virtualMachineScaleSet.Identity.Type = compute.ResourceIdentityTypeSystemAssignedUserAssigned
+				virtualMachineScaleSet.Identity.UserAssignedIdentities = make(map[string]*compute.VirtualMachineScaleSetIdentityUserAssignedIdentitiesValue)
+			} else if virtualMachineScaleSet.Identity.Type == compute.ResourceIdentityTypeNone {
+				// Note: in current AKS-Engine's implementation, we will never enter into this branch. Just handle it here in case implementation
```

Comment:
late comment: why not return error in this case?

Reply:
at least we should log some message?

Reply:
Yeah, logging a warning here is better, but there is no logger available in current context. As for return an error, as the function name `associateAddonIdentitiesToVMSS` indicates, this function should associate identities to VMSS, if `virtualMachineScaleSet.Identity.Type` is `None`(which will never happen in current AKS-Engine's implementation but may happen later), instead of returning error, just change `virtualMachineScaleSet.Identity.Type` to `UserAssigned` and associate addons' identities to the VMSS makes more sense.